README.md
hook-library
============

API Hook Library
(C) 2010 - Guilherme Maeda

This is a free library for hooking any exported function from a Windows
executable file.

It uses the JMP replacement method for detouring the function calls, and it
also uses a disassembler (ollyasm) to correctly handle functions in any calling
convention (cdecl, stdcall, fastcall, etc), as long as the entire function code
is at least 5 bytes long.

It was compiled and tested using Pelles C on Windows XP and Windows 7.

Please check the example and source files for instruction on how to use the
library.

CHANGELOG
Version 1.0 - May 05, 2014
Documented the code, included an usage example and added the library to github.


src/hook.c
// API Hook Library -- Hook functions
//
// Copyright (C) 2010 Guilherme Maeda
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#include <windows.h>
#include "ollyasm\disasm.h"
#include "hook.h"

BOOL HookInstall (const char * szModule, const char * szFunction, void * pStub, void ** pHopFunction, PHOOKDATA pHookData) {
	DWORD iProtect, iUseless;

	// Get the function address //
	pHookData->pFunction = (unsigned char *) GetProcAddress (GetModuleHandle (szModule), szFunction);
	if (! pHookData->pFunction) return FALSE;

	// Determine the hook size (by disassembling the original function)
	if (! HookGetSize (pHookData))
		return FALSE;

	// Unprotect the first X bytes //
	if (! VirtualProtect (pHookData->pFunction, pHookData->iHookSize, PAGE_EXECUTE_READWRITE, &iProtect))
		return FALSE;

	// Backup the first X bytes //
	memcpy (pHookData->szBackup, pHookData->pFunction, pHookData->iHookSize);

	// Set the first 5 bytes to [JMP pStub] and the rest to NOPs //
	pHookData->pFunction[0] = 0xE9;
	*(DWORD *) &pHookData->pFunction[1] =
		(DWORD) pStub - (DWORD) pHookData->pFunction - 5;
	for (int i=5; i<pHookData->iHookSize; i++)
		pHookData->pFunction[i] = 0x90;

	// Reprotect the first X bytes //
	VirtualProtect (pHookData->pFunction, pHookData->iHookSize, iProtect, &iUseless);

	// Set the hook hop function //
	*(DWORD *) pHopFunction = (DWORD) pHookData->szBackup;
	VirtualProtect (*pHopFunction, pHookData->iHookSize, PAGE_EXECUTE_READWRITE, &iUseless);

	// Set the hook hop helper //
	pHookData->szBackup[pHookData->iHookSize] = 0xE9;
	*(DWORD *) &(pHookData->szBackup[pHookData->iHookSize+1]) =
		(DWORD) pHookData->pFunction - (DWORD) pHookData->szBackup - 5;
	VirtualProtect (pHookData->szBackup, HOOK_BACKUP_SIZE, PAGE_EXECUTE_READWRITE, &iUseless);

	return TRUE;
}

BOOL HookUninstall (PHOOKDATA pHookData) {
	DWORD iProtect, iUseless;

	// Unprotect the first X bytes //
	if (! VirtualProtect (pHookData->pFunction, pHookData->iHookSize, PAGE_EXECUTE_READWRITE, &iProtect))
		return FALSE;

	// Restore the first X bytes from the backup //
	memcpy (pHookData->pFunction, pHookData->szBackup, pHookData->iHookSize);

	// Reprotect the first X bytes //
	VirtualProtect (pHookData->pFunction, pHookData->iHookSize, iProtect, &iUseless);

	return TRUE;
}


BOOL HookGetSize (PHOOKDATA pHookData) {
	t_disasm stDa;
	unsigned char * pCurOp;
	int iOpSize;

	// Disassemble the beginning of the hooked function to determine the correct
	// size to be changed, in order not to have the hop land in the middle of
	// an operand.
	pCurOp = pHookData->pFunction;
	pHookData->iHookSize = 0;
	while (pHookData->iHookSize < 5) { // 5 is the minimum for the JMP
		iOpSize = Disasm(pCurOp, MAXCMDSIZE, 0, &stDa, DISASM_SIZE);
		if (! iOpSize)
			return FALSE;

		pCurOp += iOpSize;
		pHookData->iHookSize += iOpSize;
	}

	return TRUE;
}


example/winmain.c
#include <stdio.h>
#include <windows.h>
#include "..\src\hook.h"

/**********************************************************************
 In this example, we'll hook the MessageBoxW function, declared as:
	int WINAPI MessageBoxW(
	  _In_opt_  HWND hWnd,
	  _In_opt_  LPCWSTR lpText,
	  _In_opt_  LPCWSTR lpCaption,
	  _In_      UINT uType
	);

**********************************************************************/

// Declare the hooks
HOOK_DECLARE(
	MyMessageBox, // The hook name (this can be anything you want)
	int,          // The return type of the hooked function
	WINAPI,       // The calling convention of the hooked function (WINAPI = stdcall)
	(HWND, LPCWSTR, LPCWSTR, UINT) // The hooked function parameters
);


BOOL WINAPI WinMain (HINSTANCE hInstance, HINSTANCE hPrevInst, char * szCmdLine, int iShow) {
	// Show the message box
	MessageBoxW(NULL, L"I have a dog named Foobar!", L"hook_example", MB_OK);

	// Install the hook
	HOOK_INSTALL(
		MyMessageBox, // Hook name
		"User32.dll", // Hooked library name
		"MessageBoxW" // Hooked function name
	);

	// Show the message box again (something different should happen now)
	MessageBoxW(NULL, L"I have a dog named Foobar!", L"hook_example", MB_OK);

	// Uninstall the hook
	HOOK_UNINSTALL(MyMessageBox);

	// Show the message box for the last time (it should be back to normal)
	MessageBoxW(NULL, L"I have a dog named Foobar!", L"hook_example", MB_OK);

	return FALSE;
}


// This is the stub function
HOOK_STUB_FUNCTION(
		MyMessageBox, // Hook name
		int,          // Return type (these must be the same as the hooked function)
		WINAPI,       // Calling convention
	  (HWND hWnd, LPCWSTR lpText, LPCWSTR lpCaption, UINT uType) // Parameters
) {
	char szText[1024];
	
	// Call the original function using a different title for the message box
	return HOOK_HOP(MyMessageBox)(hWnd, lpText, L"This call was hijacked!", uType);
}


src/hook.h
// API Hook Library -- Header file
//
// Copyright (C) 2010 Guilherme Maeda
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#ifndef HOOK_H_INCLUDED
#define HOOK_H_INCLUDED

/***********************************************************************************\
* How the hook works:                                                               *
*  __________       __________       __________       __________                    *
* |          |---->||  Jump  ||---->|          |---->| Original |                   *
* |          |     ||________||     |          |     | X bytes  |                   *
* |          |     |          |<---------------------|__________|                   *
* | Calling  |     |  Hooked  |     |   Stub   |       Hook hop                     *
* | Process  |     | Function |     | Function |        helper                      *
* |          |     |__________|---->|          |                                    *
* |          |                      |          |                                    *
* |__________|<---------------------|__________|                                    *
*                                                                                   *
* 1) The process calls the hooked function.                                         *
* 2) The hooked function jumps to the stub function.                                *
* 3) The stub function calls the hooked function's original first bytes.            *
* 4) The first bytes jump to the hooked function right after the hook. (Hook hop)   *
* 5) After the hooked function returns, the stub function returns to the process.   *
\***********************************************************************************/

/*
 * Macro to declare the necessary objects for a hook:
 *  name: The hook name (must be a valid C variable/function name).
 *  ret_type: Return type of the hooked function.
 *  call_conv: Calling convention of the hooked function.
 *  params: The function parameters (enclosed in parenthesis).
 *
 * Usage example: HOOK_DECLARE(MyMessageBeep, BOOL, WINAPI, (UINT));
 */
#define HOOK_DECLARE(name, ret_type, call_conv, params) \
	ret_type call_conv name##_Stub params; \
	static ret_type (call_conv * name##_Hop) params; \
	static HOOKDATA name##_HookData;

/*
 * Macro to install/enable a hook:
 *  name: The hook name (must be the same name used in HOOK_DECLARE).
 *  lib: The DLL library name.
 *  expName: The hooked function's exported name in the library.
 *
 * Usage example: INSTALL_HOOK(MyMessageBeep, "User32.dll", "MessageBeep");
 */
#define HOOK_INSTALL(name, lib, expName) \
	HookInstall(lib, expName, (void *) name##_Stub, (void **) &name##_Hop, &name##_HookData)

/*
 * Macro to uninstall/disable a hook:
 *  name: The hook name (must be the same name used in HOOK_DECLARE).
 *
 * Usage example: HOOK_UNINSTALL(MyMessageBeep);
 */
#define HOOK_UNINSTALL(name) \
	HookUninstall(&name##_HookData)

/*
 * Macro to declare the Stub function implementation:
 *  name: The hook name (must be the same name used in HOOK_DECLARE).
 *  ret_type: Return type of the hooked function (eg DWORD).
 *  call_conv: Calling convention of the hooked function (eg __stdcall).
 *  params: The function parameters (enclosed in parenthesis).
 *
 * Usage example:
 *  HOOK_STUB_FUNCTION(MyMessageBeep, BOOL, WINAPI, (UINT uType)) {
 *      // Make every beep sound like an error lol
 *      return HOOK_HOP(MyMessageBeep) (MB_ICONERROR);
 *  }
 */
#define HOOK_STUB_FUNCTION(name, ret_type, call_conv, params) \
	ret_type call_conv name##_Stub params

/*
 * Macro to use the Hop function:
 *  name: The hook name (must be the same name used in HOOK_DECLARE).
 *
 * Usage example: HOOK_HOP(MyMessageBeep)(MB_ICONERROR);
 */
#define HOOK_HOP(name) name##_Hop


#ifndef __thiscall
#define __thiscall __stdcall
#endif // __thiscall

#define HOOK_BACKUP_SIZE 32

// Hook control structure
typedef struct _HOOKDATA {
	unsigned char * pFunction;
	void ** pHopFunction;
	DWORD iHopAddress;
	int iHookSize;
	unsigned char szBackup[HOOK_BACKUP_SIZE];
} HOOKDATA, * PHOOKDATA;


// API functions
BOOL HookInstall (const char * szModule, const char * szFunction, void * pStub, void ** pHopFunction, PHOOKDATA pHookData);
BOOL HookUninstall (PHOOKDATA pHookData);

// Internal functions
BOOL HookGetSize (PHOOKDATA pHookData);


#endif // HOOK_H_INCLUDED
